File: Exercise_2.cpp
/* 
    Time Complexity                              :  O(N)
    Space Complexity                             :  O(1)
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  

// https://leetcode.com/problems/recover-binary-search-tree/

class Solution {
private:
    bool fixed;
    TreeNode *x, *y;
    TreeNode *prev;
public:
    void recoverTree(TreeNode* root) {
        inOrder(root);
        swap(x->val,y->val);
    }
    
    void inOrder(TreeNode *root) {
        if(!root or fixed == true) return;
        
        inOrder(root->left);
        if(prev and root->val < prev->val) {
            if(!x) {
                x = prev;
                y = root;
            } else {
                y = root;
            }
        }
        
        prev = root;
        inOrder(root->right);
        
    }
};

File: Exercise_2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode t1, t2, prev;
    public void recoverTree(TreeNode root) {
        t1 = t2 = prev = null;

        inorder(root);

        int temp = t1.val;
        t1.val = t2.val;
        t2.val = temp;
    }

    private void inorder(TreeNode curr){
        if(curr != null){
            inorder(curr.left);

            if(prev!=null && curr.val < prev.val){
                if(t1 == null){
                    t1 = prev;
                }
                t2 = curr;
            }

            prev = curr;

            inorder(curr.right);
        }
    }
}

//tc = O(n)
//sc = O(n)
